### PR TITLE
 TFT model: Model class implementation

### PR DESCRIPTION
**Pull request by egoriyaa**
## Before submitting (must do checklist)

- [ ] Did you read the [contribution guide](https://github.com/etna-team/etna/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs? We use Numpy format for all the methods and classes.
- [ ] Did you write any new necessary tests?
- [ ] Did you update the [CHANGELOG](https://github.com/etna-team/etna/blob/master/CHANGELOG.md)?

### CODE REVIEW

File: etna/models/nn/tft_native/loss.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        pred:
+            pred values.
+
+        Returns
+        -------
+        :
+            loss
+
+        """
+        return inputs + pred
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/loss.py:28]: Not actually looks like quantile loss

File: etna/models/nn/tft_native/tft.py [hunk 2]
```diff
@@ -0,0 +1,691 @@
...
+    from etna.models.nn.tft_native.layers import TemporalFusionDecoder
+    from etna.models.nn.tft_native.layers import VariableSelectionNetwork
+    from etna.models.nn.tft_native.loss import QuantileLoss
+
+
+class TFTNativeBatch(TypedDict):
+    """Batch specification for TFT."""
+
+    decoder_target: "torch.Tensor"
+    static_reals: Dict[str, "torch.Tensor"]
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:33]: Is it efficient way to store dict of tensors? Can't we avoid using dicts?

File: etna/models/nn/tft_native/tft.py [hunk 3]
```diff
@@ -0,0 +1,691 @@
...
+    ----
+    This model requires ``torch`` extension to be installed.
+    Read more about this at :ref:`installation page <installation>`.
+    """
+
+    def __init__(
+        self,
+        encoder_length: int,
+        decoder_length: int,
+        n_heads: int = 2,
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:558]: How did you choose the defaults?

File: etna/models/nn/tft_native/tft.py [hunk 4]
```diff
@@ -0,0 +1,691 @@
...
+        num_layers: int = 2,
+        dropout: float = 0.0,
+        hidden_size: int = 16,
+        lr: float = 1e-3,
+        static_categoricals: List[str] = [],
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:563]: Default values should't be mutable objects, linters warn you about that

File: etna/models/nn/tft_native/tft.py [hunk 5]
```diff
@@ -0,0 +1,691 @@
...
+        static_reals: List[str] = [],
+        time_varying_categoricals_encoder: List[str] = [],
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:565]: Why do we have separate lists for encoder and decoder? It is for the case when feature is known/unknown in the future?

Comment (alex-hse-repository) [reply etna/models/nn/tft_native/tft.py:565]: Can't we define them automatically based on metadata of TSDataset? We know which columns are regressors + which are categoricals

File: etna/models/nn/tft_native/tft.py [hunk 6]
```diff
@@ -0,0 +1,691 @@
...
+        time_varying_categoricals_decoder: List[str] = [],
+        time_varying_reals_encoder: List[str] = ["target"],
+        time_varying_reals_decoder: List[str] = ["target"],
+        categorical_feature_to_id: Dict[str, Dict[str, int]] = {},
+        loss: Optional["QuantileLoss"] = None,
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:570]: Which loss can we use instead? May be we need base class for possible losses

File: etna/models/nn/tft_native/tft.py [hunk 7]
```diff
@@ -0,0 +1,691 @@
...
+        dropout: float = 0.0,
+        hidden_size: int = 16,
+        lr: float = 1e-3,
+        static_categoricals: List[str] = [],
+        static_reals: List[str] = [],
+        time_varying_categoricals_encoder: List[str] = [],
+        time_varying_categoricals_decoder: List[str] = [],
+        time_varying_reals_encoder: List[str] = ["target"],
+        time_varying_reals_decoder: List[str] = ["target"],
+        categorical_feature_to_id: Dict[str, Dict[str, int]] = {},
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:569]: What do we need it for?

File: etna/models/nn/tft_native/tft.py [hunk 8]
```diff
@@ -0,0 +1,691 @@
...
+                lr=lr,
+                static_categoricals=static_categoricals,
+                static_reals=static_reals,
+                time_varying_categoricals_encoder=time_varying_categoricals_encoder,
+                time_varying_categoricals_decoder=time_varying_categoricals_decoder,
+                time_varying_reals_encoder=time_varying_reals_encoder,
+                time_varying_reals_decoder=time_varying_reals_decoder,
+                categorical_feature_to_id=categorical_feature_to_id,
+                optimizer_params=optimizer_params,
+                loss=QuantileLoss() if loss is None else loss,
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:660]: Better to do it in current constructor
```
 self.loss = QuantileLoss() if loss is None else loss 
```

File: etna/models/nn/tft_native/tft.py [hunk 9]
```diff
@@ -0,0 +1,691 @@
...
+            test_dataloader_params=test_dataloader_params,
+            val_dataloader_params=val_dataloader_params,
+            trainer_params=trainer_params,
+            split_params=split_params,
+        )
+
+    def params_to_tune(self) -> Dict[str, BaseDistribution]:
+        """Get default grid for tuning hyperparameters.
+
+        This grid tunes parameters: ``num_layers``, ``hidden_size``, ``lr``, ``encoder_length``.
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:676]: It is not true, you tune more parameters

File: etna/models/nn/tft_native/tft.py [hunk 10]
```diff
@@ -0,0 +1,691 @@
...
+        Other parameters are expected to be set by the user.
+
+        Returns
+        -------
+        :
+            Grid to tune.
+        """
+        return {
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:684]: Why did you choose these parameters? In the paper the set was a little bit different

File: etna/models/nn/tft_native/tft.py [hunk 11]
```diff
@@ -0,0 +1,691 @@
...
+        encoder_length:
+            encoder length
+        decoder_length:
+            decoder length
+        n_heads:
+            number of heads in Multi-Head Attention
+        num_layers:
+            number of layers in LSTM layer
+        dropout:
+            dropout rate in rnn layer
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:76]: We actually have dropout not only in rnn

File: etna/models/nn/tft_native/tft.py [hunk 12]
```diff
@@ -0,0 +1,691 @@
...
+        self.time_varying_reals_encoder = time_varying_reals_encoder
+        self.time_varying_reals_decoder = time_varying_reals_decoder
+        self.categorical_feature_to_id = categorical_feature_to_id
+        self.loss = loss
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+
+        if self.static_categoricals:
+            self.static_embeddings = nn.ModuleDict(
+                {
+                    feature: nn.Embedding(len(self.categorical_feature_to_id[feature]), self.hidden_size)
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:122]: Why `num_embeddings = len(self.categorical_feature_to_id[feature])`?

File: etna/models/nn/tft_native/tft.py [hunk 13]
```diff
@@ -0,0 +1,691 @@
...
+                    for feature in self.static_categoricals
+                }
+            )
+        if self.static_reals:
+            self.static_scalers = nn.ModuleDict(
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:127]: Why we can't just use `nn.Linear(n_static_real_features, self.hidden_size)`?

File: etna/models/nn/tft_native/tft.py [hunk 14]
```diff
@@ -0,0 +1,691 @@
...
+        x:
+            batch of data
+
+        Returns
+        -------
+        :
+            transformed batch of data
+        """
+        # Apply transformation to static data
+        if self.static_reals:
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:233]: It is more explicit to write sth like
```
if self.static_reals is not None
```

File: etna/models/nn/tft_native/tft.py [hunk 15]
```diff
@@ -0,0 +1,691 @@
...
+                    for feature in self.static_categoricals
+                }
+            )
+        if self.static_reals:
+            self.static_scalers = nn.ModuleDict(
+                {feature: nn.Linear(1, self.hidden_size) for feature in self.static_reals}
+            )
+
+        if self.time_varying_categoricals_encoder:
+            self.time_varying_embeddings_encoder = nn.ModuleDict(
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:132]: Better do define all the attributes with `None` and then assigne them to some values -- not this attribute might not exists in some cases

File: etna/models/nn/tft_native/tft.py [hunk 16]
```diff
@@ -0,0 +1,691 @@
...
+        self.static_categoricals = static_categoricals
+        self.static_reals = static_reals
+        self.time_varying_categoricals_encoder = time_varying_categoricals_encoder
+        self.time_varying_categoricals_decoder = time_varying_categoricals_decoder
+        self.time_varying_reals_encoder = time_varying_reals_encoder
+        self.time_varying_reals_decoder = time_varying_reals_decoder
+        self.categorical_feature_to_id = categorical_feature_to_id
+        self.loss = loss
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:118]: Why do we need so much "reshapings"? Can't we just define dict of embedding layers for  categoricals and one linear layer for reals?

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/models/nn/tft_native/tft.py [hunk 17]
```diff
@@ -0,0 +1,691 @@
...
+    from etna.models.nn.tft_native.layers import StaticCovariateEncoder
+    from etna.models.nn.tft_native.layers import TemporalFusionDecoder
+    from etna.models.nn.tft_native.layers import VariableSelectionNetwork
+    from etna.models.nn.tft_native.loss import QuantileLoss
+
+
+class TFTNativeBatch(TypedDict):
+    """Batch specification for TFT."""
+
+    decoder_target: "torch.Tensor"
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:32]: Is `encoder_target` included in `time_varying_reals_encoder`?

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

File: etna/models/nn/tft_native/tft.py (hunk 6 above)
Comment (egoriyaa) [reply etna/models/nn/tft_native/tft.py:570]: By default nn.MSE, additionally implement QuantileLoss

File: etna/models/nn/tft_native/tft.py (hunk 12 above)
Comment (egoriyaa) [reply etna/models/nn/tft_native/tft.py:122]: Instead additional dict with num of unique values for feature

File: etna/models/nn/tft_native/tft.py [hunk 18]
```diff
@@ -0,0 +1,691 @@
...
+        num_layers:
+            number of layers in LSTM layer
+        dropout:
+            dropout rate in rnn layer
+        hidden_size:
+            size of the hidden state
+        lr:
+            learning rate
+        static_categoricals:
+            categorical features for the whole series, e.g. `segment`
```
Comment (d-a-bunin) [inline etna/models/nn/tft_native/tft.py:82]: How are we going to control static features? In `TSDataset` we doens't have any flags about static features.

Comment (egoriyaa) [reply etna/models/nn/tft_native/tft.py:81]: What means "control"? User is responsible for his inputs

**Comment by egoriyaa** [comment]
There was an idea to reimplement `fit` method of base class to fill feature names if user passes nothing.
I found it strange(and overloaded) from the code point of view to classify all the columns of the dataset.

I can try again when this version will be ok

File: etna/models/nn/tft_native/tft.py [hunk 19]
```diff
@@ -683,9 +645,9 @@ def params_to_tune(self) -> Dict[str, BaseDistribution]:
...
-            "n_heads": IntDistribution(low=1, high=3),
-            "hidden_size": IntDistribution(low=4, high=64, step=4),
-            "lr": FloatDistribution(low=1e-5, high=1e-2, log=True),
-            "encoder_length": IntDistribution(low=1, high=20),
-            "dropout": FloatDistribution(low=0.0, high=0.3),
+            "n_heads": CategoricalDistribution([1, 4]),
+            "hidden_size": CategoricalDistribution([16, 20, 40, 80, 160, 240, 320]),
+            "lr": FloatDistribution(low=1e-4, high=1e-2, log=True),
+            "dropout": FloatDistribution(low=0.1, high=0.9, step=0.1),
+            "train_batch_size": IntDistribution(low=64, high=256, log=True),
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:652]: First of all, it was categorical distribution in the paper, but do we really need to optimize it? In other nns we don't perform batch_size optimisation

File: etna/models/nn/tft_native/tft.py [hunk 20]
```diff
@@ -73,7 +72,7 @@ def __init__(
         num_layers:
             number of layers in LSTM layer
         dropout:
-            dropout rate in rnn layer
+            dropout rate
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:75]: You should fix it in model constructor too

File: etna/models/nn/tft_native/tft.py [hunk 21]
```diff
@@ -87,9 +86,9 @@ def __init__(
         time_varying_categoricals_decoder:
             time varying categorical features for decoder (known for future)
         time_varying_reals_encoder:
-            time varying continuous features for encoder, default to `target` with lag 1
+            time varying continuous features for encoder, default to `target`
         time_varying_reals_decoder:
-            time varying continuous features for decoder (known for future), default to `target` with lag 1
+            time varying continuous features for decoder (known for future), default to `target`
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:91]: We don't know target in the future, so it is empty by default

File: etna/models/nn/tft_native/tft.py [hunk 22]
```diff
@@ -453,31 +385,45 @@ def step(self, batch: TFTNativeBatch, *args, **kwargs):  # type: ignore
...
         def _make(
-            values_target: np.ndarray,
             df: pd.DataFrame,
             start_idx: int,
             encoder_length: int,
             decoder_length: int,
         ) -> Optional[dict]:
 
             sample: Dict[str, Any] = {
+                "segment": None,
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:426]: Where do you use it? If you use it somewhere you should add it to `TFTNativeBatch` definition

File: tests/test_models/test_nn/tft_native/test_tft_native.py [hunk 23]
```diff
@@ -0,0 +1,95 @@
...
+            encoder_length=encoder_length,
+            decoder_length=decoder_length,
+        )
+    )
+    first_sample = ts_samples[0]
+
+    assert first_sample["segment"] == "segment_1"
+    assert first_sample["decoder_target"].shape == (decoder_length, 1)
+
+    assert len(first_sample["static_reals"]) == 0
```
Comment (alex-hse-repository) [inline tests/test_models/test_nn/tft_native/test_tft_native.py:62]: Better to create test checking all the cases of batch

File: tests/test_models/test_nn/tft_native/test_tft_native.py [hunk 24]
```diff
@@ -0,0 +1,95 @@
...
+        first_sample["time_varying_reals_encoder"]["target"],
+    )
+
+
+@pytest.mark.parametrize("encoder_length", [1, 2, 10])
+def test_context_size(encoder_length):
+    encoder_length = encoder_length
+    decoder_length = encoder_length
+    model = TFTNativeModel(
+        encoder_length=encoder_length, decoder_length=decoder_length, trainer_params=dict(max_epochs=100)
```
Comment (alex-hse-repository) [inline tests/test_models/test_nn/tft_native/test_tft_native.py:80]: Remove `trainer_params` -- it is not necessary for this test, be minimalistic)

File: tests/test_models/test_nn/tft_native/test_tft_native.py [hunk 25]
```diff
@@ -0,0 +1,95 @@
+import numpy as np
```
Comment (alex-hse-repository) [inline tests/test_models/test_nn/tft_native/test_tft_native.py:1]: May be we can also add test for `backtest`

File: tests/test_models/test_nn/tft_native/test_tft_native.py (hunk 23 above)
Comment (alex-hse-repository) [reply tests/test_models/test_nn/tft_native/test_tft_native.py:62]: Also it is better to set all parameters with different column types in the other tests as the affects the architecture, some steps are not performed if they are unset

File: etna/models/nn/tft_native/tft.py [hunk 26]
```diff
@@ -0,0 +1,653 @@
...
+        """
+        target_pred = self.forward(batch)  # (batch_size, decoder_length, 1)
+        target_true = batch["decoder_target"].float()  # (batch_size, decoder_length, 1)
+        loss = self.loss(target_pred, target_true)
+        return loss, target_true, target_pred
+
+    def make_samples(self, df: pd.DataFrame, encoder_length: int, decoder_length: int) -> Iterator[dict]:
+        """Make samples from segment DataFrame."""
+        for feature in self.categorical_feature_to_id:
+            df[feature] = df[feature].map(self.categorical_feature_to_id[feature])
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:416]: We cat  overload `fit` at least to get this mapping `categorical_feature_to_id`

File: etna/models/nn/tft_native/tft.py [hunk 27]
```diff
@@ -0,0 +1,653 @@
...
+    def __init__(
+        self,
+        encoder_length: int,
+        decoder_length: int,
+        n_heads: int,
+        num_layers: int,
+        dropout: float,
+        hidden_size: int,
+        lr: float,
+        static_categoricals: List,
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:52]: `List[str]`

File: etna/models/nn/tft_native/tft.py [hunk 28]
```diff
@@ -0,0 +1,653 @@
...
+        self.static_reals = static_reals
+        self.time_varying_categoricals_encoder = time_varying_categoricals_encoder
+        self.time_varying_categoricals_decoder = time_varying_categoricals_decoder
+        self.time_varying_reals_encoder = time_varying_reals_encoder
+        self.time_varying_reals_decoder = time_varying_reals_decoder
+        self.categorical_feature_to_id = categorical_feature_to_id
+        self.loss = loss
+        self.optimizer_params = {} if optimizer_params is None else optimizer_params
+
+        self.num_embeddings_per_feature = self._num_embeddings_per_feature()
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:118]: Do we really need the separate method? Why can't we just place the implementation here?

File: etna/models/nn/tft_native/tft.py [hunk 29]
```diff
@@ -0,0 +1,653 @@
...
+        )
+
+        self.time_varying_scalers_encoder = nn.ModuleDict(
+            {feature: nn.Linear(1, self.hidden_size) for feature in self.time_varying_reals_encoder}
+        )
+        self.time_varying_scalers_decoder = nn.ModuleDict(
+            {feature: nn.Linear(1, self.hidden_size) for feature in self.time_varying_reals_decoder}
+        )
+
+        if self.num_static > 0:
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:151]: Better to place this assignments just after the definition

File: etna/models/nn/tft_native/tft.py [hunk 30]
```diff
@@ -0,0 +1,653 @@
...
+
+        Returns
+        -------
+        :
+            dict where keys are features and values are number of unique values for each feature.
+        """
+        return {feature: len(self.categorical_feature_to_id[feature]) for feature in self.categorical_feature_to_id}
+
+    @property
+    def num_timestamps(self) -> int:
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:218]: Should all these properties be public?

File: etna/models/nn/tft_native/tft.py [hunk 31]
```diff
@@ -0,0 +1,653 @@
...
+        """Get number of features in decoder.
+
+        Returns
+        -------
+        :
+            number of features in decoder.
+        """
+        return len(self.time_varying_reals_decoder + self.time_varying_categoricals_decoder)
+
+    def _transform_features(self, x: TFTNativeBatch):
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:261]: Add return typing

File: etna/models/nn/tft_native/tft.py [hunk 32]
```diff
@@ -0,0 +1,653 @@
...
+            x["time_varying_reals_encoder"][feature] = self.time_varying_scalers_encoder[feature](
+                x["time_varying_reals_encoder"][feature].float()
+            )  # (batch_size, encoder_length, hidden_size)
+        for feature in self.time_varying_reals_decoder:
+            x["time_varying_reals_decoder"][feature] = self.time_varying_scalers_decoder[feature](
+                x["time_varying_reals_decoder"][feature].float()
+            )  # (batch_size, decoder_length, hidden_size)
+        return x
+
+    def forward(self, x: TFTNativeBatch, *args, **kwargs):
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:304]: Add return typing

File: etna/models/nn/tft_native/tft.py [hunk 33]
```diff
@@ -0,0 +1,653 @@
...
+        """
+        target_true = x["decoder_target"].float()  # (batch_size, decoder_length, 1)
+        batch_size = target_true.size()[0]
+        batch = self._transform_features(x)
+
+        #  Pass static data through variable selection and covariate encoder blocks
+        if self.num_static > 0:
+            static_features = batch["static_reals"].copy()
+            static_features.update(batch["static_categoricals"])
+            static_features = self.static_variable_selection(static_features)  # type: ignore
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:325]: Why do you need to ignore type here? I guess you can write sth like this to solve it:
```
if self. static_variable_selection is not None and  self.static_covariate_encoder is not None:
```

File: etna/models/nn/tft_native/tft.py [hunk 34]
```diff
@@ -0,0 +1,653 @@
...
+                decoder_features = self.decoder_variable_selection(
+                    x=decoder_features, context=c_s.expand(batch_size, self.decoder_length, self.hidden_size)
+                )  # type: ignore
+                # (batch_size, decoder_length, hidden_size)
+            else:
+                decoder_features = self.decoder_variable_selection(x=decoder_features)  # type: ignore
+                # (batch_size, decoder_length, hidden_size)
+        else:
+            decoder_features = torch.zeros(batch_size, self.decoder_length, self.hidden_size)
+        residual = torch.cat((encoder_features, decoder_features), dim=1)  # type: ignore
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:355]: Why so much " # type: ignore"????

File: etna/models/nn/tft_native/tft.py [hunk 35]
```diff
@@ -0,0 +1,653 @@
...
+        )  # (batch_size, decoder_length, hidden_size)
+
+        # Pass common data through gated layer
+        features = torch.cat((encoder_features, decoder_features), dim=1)  # type: ignore
+        features = self.gated_norm1(x=features, residual=residual)  # (batch_size, num_timestamps, hidden_size)
+
+        residual = features
+
+        # Pass common data through temporal fusion block
+        if self.num_static > 0:
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:379]: Looks like we need to come up with some dummy layers to avoid to much ifs

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]
Looks good, just some minor fixes 👍

File: etna/models/nn/tft_native/tft.py (hunk 20 above)
Comment (egoriyaa) [reply etna/models/nn/tft_native/tft.py:75]: in last commit it is ok

File: tests/test_models/test_nn/tft_native/test_tft_native.py [hunk 36]
```diff
@@ -0,0 +1,238 @@
...
+    assert len(first_sample["time_varying_reals_decoder"]) == len(time_varying_reals_decoder)
+
+    for feature in static_reals:
+        np.testing.assert_almost_equal(
+            df[[feature]].iloc[:1],
+            first_sample["static_reals"][feature],
+        )
+    for feature in static_categoricals:
+        np.testing.assert_almost_equal(
+            df[[feature]].iloc[:1],
```
[the remark this answers is not in the document]
Comment (egoriyaa) [reply tests/test_models/test_nn/tft_native/test_tft_native.py:176]: .values[:1]?
arrays need to be 2D, we can do without reshape

File: etna/models/nn/tft_native/tft.py [hunk 37]
```diff
@@ -459,7 +461,8 @@ def configure_optimizers(self) -> "torch.optim.Optimizer":
 class TFTNativeModel(DeepBaseModel):
     """TFT model. For more details read the `paper <https://arxiv.org/abs/1912.09363>`_.
 
-    Model needs label encoded inputs, for that purposes use :py:class:`~etna.transforms.LabelEncoderTransform`
+    Model needs label encoded inputs, for that purposes use :py:class:`~etna.transforms.LabelEncoderTransform`.
+    Feature values that were not seen during `fit` should be set to NaN for expected behaviour
```
Comment (alex-hse-repository) [inline etna/models/nn/tft_native/tft.py:465]: Better to suggest the correct strategy for `LabelEncoderTransform `

**Review by alex-hse-repository — CHANGES_REQUESTED** [review]

**Pull request merged**